src-tauri/src/commands/mod.rs

```diff
@@ -1,36 +1,96 @@
-use crate::structs::ConnectionState;
-use citadel_internal_service_types::InternalServiceRequest;
+use crate::structs::{ConnectionState, PacketHandle};
+use citadel_internal_service_types::{InternalServiceRequest, InternalServiceResponse};
 use futures::SinkExt;
 use tauri::State;
+use tokio::sync::mpsc;
+use uuid::Uuid;
 
-pub mod connect;
-pub mod disconnect;
-pub mod get_session;
-pub mod list_all_peers;
-pub mod list_registered_peers;
-pub mod local_db_clear_all_kv;
-pub mod local_db_delete_kv;
-pub mod local_db_get_all_kv;
-pub mod local_db_get_kv;
-pub mod local_db_set_kv;
-pub mod message;
-pub mod open_connection;
+mod connect;
+// pub mod disconnect;
+// pub mod get_session;
+mod list_all_peers;
+// pub mod list_registered_peers;
+// pub mod local_db_clear_all_kv;
+// pub mod local_db_delete_kv;
+// pub mod local_db_get_all_kv;
+// pub mod local_db_get_kv;
+// pub mod local_db_set_kv;
+// pub mod message;
+// pub mod open_connection;
 pub mod peer_connect;
-pub mod peer_disconnect;
-pub mod peer_register;
-pub mod register;
-
-pub(crate) async fn send_to_internal_service(
-    request: InternalServiceRequest,
-    state: State<'_, ConnectionState>,
-) -> Result<(), String> {
-    state
-        .sink
-        .lock()
-        .await
-        .as_mut()
-        .ok_or("No connection to the internal service set")?
-        .send(request)
+// pub mod peer_disconnect;
+// pub mod peer_register;
+mod list_known_servers;
+pub mod register; // this can go private again after RegistrationRequestTS is reformatted
+
+pub use connect::connect;
+pub use list_all_peers::list_all_peers;
+pub use list_known_servers::list_known_servers;
+pub use peer_connect::peer_connect;
+pub use register::register;
+
+pub(crate) async fn send_and_recv(
+    payload: InternalServiceRequest,
+    request_id: Uuid,
+    state: &State<'_, ConnectionState>,
+) -> InternalServiceResponse {
+    // Send message to internal service
+    println!(
+        "Sending message with request_id {}:\n{:#?}",
+        request_id, payload
+    );
+    let mut guard = state.sink.lock().await;
+    guard
+        .send(payload)
         .await
         .map_err(|err| err.to_string())
+        .expect("error sending payload to stream");
+    drop(guard);
+
+    // Create a new mpsc channel and attach the request id to it
+    let (tx, mut rx) = mpsc::channel::<InternalServiceResponse>(1024);
+    let packet_handle = PacketHandle {
+        request_id,
+        channel: tx,
+    };
+
+    // Attach the mpsc channel to the vector of listeners
+    // NOTE: be careful touching this; very easy to end up in a deadlock
+    let mut guard = state.listeners.lock().await;
+    guard.push(packet_handle);
+    drop(guard);
+
+    // Wait for the background TCP listener (main.rs) to dispatch the message
+    let incoming = match rx.recv().await {
+        Some(v) => v,
+        None => panic!("Channel unexpectedly closed before response."),
+    };
+
```

Comment:
Store the packet handle *first*. Otherwise, you allow the possibility for this scenario:

1. You send a packet
2. The IS responds almost instantly
3. the inbound packet gets routed, but since we haven't stored the transmitter yet, the packet has nowhere to go
4. Then, you register the packet_handle

It's not clear how we're handling packets that passively arrive. If this code works for now, great, however, this will all change once the PR is merged: https://github.com/Avarok-Cybersecurity/citadel-internal-service/pull/40

Your logic will simplify greatly, as you will just use the new Messenger layer.